File: goods-partner-backend/src/main/java/com/goodspartner/entity/Route.java
package com.goodspartner.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.OrderBy;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

@Getter
@Setter
@Entity
@NoArgsConstructor
@Table(name = "routes")
public class Route {

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "routes_sequence")
    @SequenceGenerator(name = "routes_sequence", sequenceName = "routes_sequence")
    private Long id;

    @Enumerated(value = EnumType.STRING)
    private RouteStatus status;
    private double totalWeight;
    private Integer totalPoints;
    private Integer totalOrders;
    private Double distance;
    private Long estimatedTime;
    private LocalDateTime startTime;
    private LocalDateTime finishTime;
    private Long spentTime; // TODO Minutes? Only for completed so far
    private Boolean optimization;

    @OneToMany(mappedBy = "route", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("completedAt ASC, expectedCompletion ASC")
    private List<RoutePoint> routePoints = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "delivery_id", referencedColumnName = "id")
    private Delivery delivery;

    @OneToOne
    private Car car;

    @OneToOne
    private Store store;

    public void setRoutePoints(List<RoutePoint> routePoints) {
        List<RoutePoint> requiredRoutePoints = Optional.ofNullable(routePoints)
                .orElse(Collections.emptyList());
        this.routePoints.clear();
        this.routePoints.addAll(requiredRoutePoints);
        this.routePoints.forEach(routePoint -> routePoint.setRoute(this));
    }
}

File: goods-partner-frontend/src/model/Route.ts
import MapPoint from './MapPoint'
import {Car} from './Car'

export interface Route {
    car: Car
    distance: number
    estimatedTime: number
    finishTime: Date
    id: number
    optimization: boolean
    routePoints: RoutePoint[]
    spentTime: number
    startTime: Date
    status: string
    store: Store
    totalOrders: number
    totalPoints: number
    totalWeight: number
}

export interface Store {
    address: string
    mapPoint: MapPoint
    name: string
}

export interface RoutePoint {
    address: string
    addressTotalWeight: number
    clientName: string
    completedAt: string
    deliveryEnd: string
    deliveryStart: string
    expectedArrival: string
    expectedCompletion: string
    id: number
    mapPoint: MapPoint
    routePointDistantTime: number
    status: RoutePointStatus
    matchingExpectedDeliveryTime: boolean
}

export enum RoutePointStatus {
    PENDING = "PENDING",
    INPROGRES = "INPROGRESS",
    DONE = "DONE",
    SKIPPED = "SKIPPED"
}
